### PR TITLE
Bugfix ensure that the entire part is matched

### PR DESCRIPTION
**Pull request by pgjones**
Checklist:

- [x] Add tests that demonstrate the correct behavior of the change. Tests should fail without the change.
- [x] Add or update relevant docs, in the docs folder and in code.
- [x] Add an entry in `CHANGES.rst` summarizing the change and linking to the issue.
- [x] Add `.. versionchanged::` entries in any relevant code docs.
- [x] Run `pre-commit` hooks and fix any issues.
- [x] Run `pytest` and `tox`, no tests failed.

### CODE REVIEW

**Comment by davidism** [comment]
Can you fix the title on this PR? Also remember to open PRs from separate branches otherwise your branch could get out of sync with ours.

**Review by kevin-brown — APPROVED** [review]
I can confirm this fixes the bug I was seeing in my local test bench related to this.

File: src/werkzeug/routing/rules.py [hunk 1]
```diff
@@ -625,7 +625,7 @@ def _parse_rule(self, rule: str) -> t.Iterable[RulePart]:
                     -len(argument_weights),
                     argument_weights,
                 )
-                if final:
+                if not static:
                     content += r"$\Z"
```
Comment (davidism) [inline src/werkzeug/routing/rules.py:629]: Shouldn't this just be `\Z`? The `$` shouldn't be needed.

Comment (pgjones) [reply src/werkzeug/routing/rules.py:629]: I think this is the reason, https://github.com/pallets/werkzeug/commit/d761d0e7acceb89f56d76434a74aeacf95d5ed49 looks like you know better :)

**Pull request merged**